internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -15,33 +15,165 @@
 package buffer
 
 import (
-	"bytes"
+	"fmt"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/logger"
 )
 
+type offset struct {
+	// [ start, end) is the range of file offset for which the
+	// data is currently stored in the buffer.
+	start, end int64
+}
+
+func (o *offset) advanceBy(value int64) {
+	o.start += value
+	o.end += value
+}
+
+type inMemoryBuffer struct {
+	buffer []byte
+	offset offset
+}
+
+// InMemoryWriteBuffer implements WriteBuffer and is stored in memory.
+// Refer to buffer.WriteBuffer for sample code usage.
 type InMemoryWriteBuffer struct {
-	// Holds the data to be written to GCS. At any time, Buffer holds the data for
-	// 2 GCS write calls.
-	buffer *bytes.Buffer
+	// Holds the incoming data from kernel write calls, currently being buffered
+	// for upload in the future.
+	current inMemoryBuffer
+
+	// Holds the data that is currently being uploaded asynchronously to GCS. Data
+	// in this buffer will persist until the entire data is successfully uploaded.
+	flushed inMemoryBuffer
+
+	// Holds the last offset to which data is written.
+	fileSize int64
+
+	// Holds the buffer's size (in bytes) to be allocated to currentBuffer and flushedBuffer.
+	bufferSize int64
 }
 
-// CreateInMemoryWriteBuffer creates a buffer with InMemoryWriteBuffer.buffer
-// set to nil. Memory is allocated to the buffer when the first write call comes.
-// This avoids unnecessarily bloating GCSFuse memory consumption.
-//
-// To allocate memory to the buffer, use WriteBuffer.InitializeBuffer.
-func CreateInMemoryWriteBuffer() *InMemoryWriteBuffer {
-	b := &InMemoryWriteBuffer{}
+// CreateInMemoryWriteBuffer returns a new InMemoryWriteBuffer.
+// Memory is lazily allocated to current and flushed buffer on their first usage.
+func CreateInMemoryWriteBuffer(bufferSizeMB uint) (*InMemoryWriteBuffer, error) {
+	if bufferSizeMB == 0 {
+		return nil, fmt.Errorf(ZeroSizeBufferError)
+	}
+	b := &InMemoryWriteBuffer{
+		bufferSize: int64(bufferSizeMB * MiB),
+	}
 	// TODO: set mtime attribute.
-	return b
+	logger.Debugf("TODO: set mtime attribute.")
+	return b, nil
+}
+
+// Allocates memory to currentBuffer if it hasn't been allocated already.
+func (b *InMemoryWriteBuffer) ensureCurrentBuffer() error {
+	if b.bufferSize == 0 {
```

Comment:
nit: This is technically not possible because CreateInMemoryWriteBuffer can't set it to zero, and that's the only approved way to create an InMemoryWriteBuffer. So, this check is really redundant.